### DISCUSSION TITLE
How to access Activity outcomes in Elsa v3 C# code workflows?

### DISCUSSION THREAD

**Discussion by kyle-alm** [elsa-workflows/elsa-core#5670] [Q&A]
Hello, I'm working on writing some workflows, and I have custom activities that check on certain True/False conditions, like one that see's if it's currently business hours / daytime. That activity has a Result output that returns the number of hours until the start of business for the case when it's not currently business hours:

`        protected override async ValueTask ExecuteAsync(ActivityExecutionContext context)
        {
            var hoursLeft = HoursUntil8am(localTime.Value);
            context.SetResult(hoursLeft);

            if (hoursLeft >= window)
            {
                await context.CompleteActivityWithOutcomesAsync("True");
            }

            await context.CompleteActivityWithOutcomesAsync("False");
        }`

What I want to know is how to access the outcomes in the workflow, so that I can branch according to outcome. I can't find any examples in the documentation or samples that show this occurring. So here:

`    public class MyWorkflow : WorkflowBase
    {
        protected override void Build(IWorkflowBuilder builder)
        {
            var parameters = builder.WithVariable<dynamic>().WithWorkflowStorage();
            var hoursUntilMorning = builder.WithVariable<int>().WithWorkflowStorage();
            var daysUntilStartdate = builder.WithVariable<int>().WithWorkflowStorage();

            builder.Root = new Sequence
            {
                Activities =
                {
                    new HttpEndpoint
                    {
                        Path = new("Messages/SponsorDailyRegistrationEmail"),
                        SupportedMethods = new(new[]{HttpMethods.Post}),
                        CanStartWorkflow = true,
                        ParsedContent = new(parameters)
                    },
                    new If
                    {
                        Condition = new(!parameters.HasProperty("MyId") || !parameters.HasProperty("Zipcode") || !parameters.HasProperty("RecipientName") || !parameters.HasProperty("RecipientAddress")),
                        Then = new Sequence
                        {
                            Activities =
                            {
                                new WriteHttpResponse
                                {
                                    ContentType = new(MediaTypeNames.Application.Json),
                                    Content = new("Payload must include Opid, Zipcode, RecipientName, and RecipientAddress.")
                                }
                            }
                        },
                        Else = new Sequence
                        {
                            Activities =
                            {
                                new IsDaytime(parameters, hoursUntilMorning)

                                // Branch if outcome is True/False

                            }
                        }
                    }
                }
            };

        }
    }`
    
    I'm starting with an Http Endpoint, then if the required parameters are passed to the endpoint, proceeding with my IsDaytime activity. I can get the output with .GetResult(), I think, but how do I know what the outcome was?

**Comment by glime-ai** [comment]
To access the outcomes of an activity in Elsa, you can use the `When` method on the builder to branch the workflow based on the outcome of the previous activity. In your case, after the `IsDaytime` activity, you would use the `When` method to handle the "True" and "False" outcomes.

Here's how you can modify your workflow to branch according to the outcome of the `IsDaytime` activity:

```csharp
public class MyWorkflow : WorkflowBase
{
    protected override void Build(IWorkflowBuilder builder)
    {
        var parameters = builder.WithVariable<dynamic>().WithWorkflowStorage();
        var hoursUntilMorning = builder.WithVariable<int>().WithWorkflowStorage();
        var daysUntilStartdate = builder.WithVariable<int>().WithWorkflowStorage();

        builder.Root = new Sequence
        {
            Activities =
            {
                new HttpEndpoint
                {
                    // ... existing setup
                },
                new If
                {
                    // ... existing setup
                    Else = new Sequence
                    {
                        Activities =
                        {
                            new IsDaytime(parameters, hoursUntilMorning)
                        }
                    }
                }
            }
        };

        // Branch based on the outcome of IsDaytime
        builder.When("True").Then<SomeActivity>(); // Replace SomeActivity with the activity you want to execute if the outcome is "True"
        builder.When("False").Then<AnotherActivity>(); // Replace AnotherActivity with the activity you want to execute if the outcome is "False"
    }
}
```

In this example, `SomeActivity` and `AnotherActivity` are placeholders for the actual activities you want to execute based on the outcome of `IsDaytime`. You would replace these with the actual activity classes you want to use.

Remember that the `When` method should be called on the same builder instance that you used to add the `IsDaytime` activity. This ensures that the workflow knows which activity's outcomes to listen for.

Please note that the `CompleteActivityWithOutcomesAsync` method in your `IsDaytime` activity should only be called once, with the correct outcome based on the condition. You should not call `CompleteActivityWithOutcomesAsync` twice as you have in your example. Instead, use an if-else statement to determine which outcome to complete with.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=dRbvTb7IhCSYiI&utm_medium=github).

**Reply by kyle-alm** [reply]
This doesn't appear to be correct for v3, or I'm missing some critical namespace information.
